src/operators/first.ts

```diff
@@ -67,100 +75,34 @@ export function first<T>(predicate: (value: T, index: number, source: Observable
-  return (source: Observable<T>) => source.lift(new FirstOperator(predicate, resultSelector, defaultValue, source));
-}
-
-class FirstOperator<T, R> implements Operator<T, R> {
-  constructor(private predicate?: (value: T, index: number, source: Observable<T>) => boolean,
-              private resultSelector?: ((value: T, index: number) => R) | void,
-              private defaultValue?: any,
-              private source?: Observable<T>) {
-  }
-
-  call(observer: Subscriber<R>, source: any): any {
-    return source.subscribe(new FirstSubscriber(observer, this.predicate, this.resultSelector, this.defaultValue, this.source));
-  }
-}
-
-/**
- * We need this JSDoc comment for affecting ESDoc.
- * @ignore
- * @extends {Ignored}
- */
-class FirstSubscriber<T, R> extends Subscriber<T> {
-  private index: number = 0;
-  private hasCompleted: boolean = false;
-  private _emitted: boolean = false;
+  let operate: OperatorFunction<T, T|R>;
 
-  constructor(destination: Subscriber<R>,
-              private predicate?: (value: T, index: number, source: Observable<T>) => boolean,
-              private resultSelector?: ((value: T, index: number) => R) | void,
-              private defaultValue?: any,
-              private source?: Observable<T>) {
-    super(destination);
-  }
-
-  protected _next(value: T): void {
-    const index = this.index++;
-    if (this.predicate) {
-      this._tryPredicate(value, index);
+  if (!predicate) {
+    if (resultSelector) {
+      operate = map(resultSelector);
     } else {
-      this._emit(value, index);
+      operate = identity;
     }
-  }
-
-  private _tryPredicate(value: T, index: number) {
-    let result: any;
-    try {
-      result = this.predicate(value, index, this.source);
-    } catch (err) {
-      this.destination.error(err);
-      return;
-    }
-    if (result) {
-      this._emit(value, index);
-    }
-  }
-
-  private _emit(value: any, index: number) {
-    if (this.resultSelector) {
-      this._tryResultSelector(value, index);
-      return;
-    }
-    this._emitFinal(value);
-  }
-
-  private _tryResultSelector(value: T, index: number) {
-    let result: any;
-    try {
-      result = (<any>this).resultSelector(value, index);
-    } catch (err) {
-      this.destination.error(err);
-      return;
+  } else {
+    if (resultSelector) {
+      operate = pipe(
+        map((value, index) => ({ value, index })),
+        filter(({ value, index }) => predicate(value, index)),
+        map(({ value, index }) => resultSelector(value, index))
+      );
+    } else {
+      operate = filter(predicate);
     }
-    this._emitFinal(result);
   }
 
-  private _emitFinal(value: any) {
-    const destination = this.destination;
-    if (!this._emitted) {
-      this._emitted = true;
-      destination.next(value);
-      destination.complete();
-      this.hasCompleted = true;
-    }
-  }
+  const after = defaultValue ? of(defaultValue) : _throw(new EmptyError());
```

Comment:
Testing whether or not `defaultValue` is falsy won't allow for default values of `0` or `false`, etc.

Reply:
Right. Don't code tired. Thanks